mid_algorithm.py

```diff
@@ -0,0 +1,76 @@
+"""
+This module contains a partially sophisticated algorithm
+"""
+
+""""""""""""""""""""""""""""""""""""""""""" Imports """""""""""""""""""""""""""""""""""""""""""
+from utils import *
+from naive_algorithm import SequenceAlgorithm
+import sklearn
+
+""""""""""""""""""""""""""""""""""""""""""" Classes """""""""""""""""""""""""""""""""""""""""""
+
+
+class MaxVarianceAlgorithm(SequenceAlgorithm):
+    """
+    A partially sophisticated algorithm that choose to buy the feature with the most variance in each stage.
+    """
+    def __init__(self, learning_algorithm: sklearn.base.ClassifierMixin):
+        """
+        Init function.
+        :param learning_algorithm: sklearn's classifier. the function saves it and uses it later.
+        """
+        super().__init__(learning_algorithm)
+        self._train_samples = None
+        self._features_costs = None
+        self._features_by_corr = None
+        self._learning_algorithm = learning_algorithm
+
+    def fit(self, train_samples: TrainSamples, features_costs: list[float]):
+        """
+        Trains the classifier. the function saves the train samples for the prediction.
+        :param train_samples: training dataset contains training data of shape (n_samples, n_features), i.e samples are
+        in the rows, and target values of shape (n_samples,). the function saves it.
+        :param features_costs: list in length number of features that contains the costs of each feature according to
+        indices. in first index you will find the cost of the first feature, etc.
+        """
+        self._train_samples = train_samples
+        self._features_costs = features_costs
+        correlations = np.var(train_samples.samples, axis=0)
+        args_sort = np.argsort(correlations)
+        self._features_by_corr = np.array(args_sort[::-1]).tolist()
+
+    def predict(self, samples: TestSamples, given_features: list[int], maximal_cost: float) -> Classes:
+        """
+        Predicts the class labels for the provided data. the function runs the fit and predict function of the given
+        sklearn's classifier on the given test samples filtering the given features only.
+        :param samples: test samples of shape (n_samples, n_features), i.e samples are in the rows.
+        :param given_features: list of the indices of the chosen features.
+        :param maximal_cost: the maximum available cost for buying features.
+        :return: Classes of shape (n_samples,) contains the class labels for each data sample.
+        """
+        given_features = self._buy_features(given_features, maximal_cost)
+        self._learning_algorithm.fit(self._train_samples.samples[:, given_features], self._train_samples.classes)
+        return self._learning_algorithm.predict(samples[:, given_features])
```

Comment:
all this function is unnecessary since SequenceAlgorithm's predict is exactly the same, so if you will remove this function, calling for MaxVarianceAlgorithm's predict will execute SequenceAlgorithm's predict and the result is the same (the code will be much shorter)